# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/test/gtest/TestQuotaManager.cpp
# Commit: d63dac1ee4d4
# Full Hash: d63dac1ee4d471defff484dae5a3535e1f8ba072
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2024-10-10 22:03:45
# Regressor Bug: 1866402
# File Overlap Count: 1
# Description:
#   Bug 1866402 - Make it possible to initialize temporary origins without ensuring origin directories; r=dom-storage-reviewers,jari
#   
#   LSNG already uses some QuotaManager APIs to achieve that origin directories are
#   not created if they don't exist during datastore preparation, but the feature
#   is not easy to use and it's also not generalized enough for use in other quota
# ==============================================================================

diff -r 2ab08c49f2b1 -r d63dac1ee4d4 dom/quota/test/gtest/TestQuotaManager.cpp
--- a/dom/quota/test/gtest/TestQuotaManager.cpp	Thu Oct 10 12:22:54 2024 +0000
+++ b/dom/quota/test/gtest/TestQuotaManager.cpp	Thu Oct 10 12:22:55 2024 +0000
@@ -23,6 +23,10 @@
   static void SetUpTestCase() { ASSERT_NO_FATAL_FAILURE(InitializeFixture()); }
 
   static void TearDownTestCase() { ASSERT_NO_FATAL_FAILURE(ShutdownFixture()); }
+
+  void TearDown() override {
+    ASSERT_NO_FATAL_FAILURE(ClearStoragesForOrigin(GetTestOriginMetadata()));
+  }
 };
 
 // Test OpenStorageDirectory when an opening of the storage directory is
@@ -1417,7 +1421,8 @@
     promises.AppendElement(quotaManager->InitializeStorage());
     promises.AppendElement(quotaManager->InitializeTemporaryStorage());
     promises.AppendElement(quotaManager->InitializeTemporaryOrigin(
-        testOriginMetadata.mPersistenceType, principalInfo));
+        testOriginMetadata.mPersistenceType, principalInfo,
+        /* aCreateIfNonExistent */ false));
 
     {
       auto value =
@@ -1436,7 +1441,8 @@
     promises.AppendElement(quotaManager->InitializeStorage());
     promises.AppendElement(quotaManager->InitializeTemporaryStorage());
     promises.AppendElement(quotaManager->InitializeTemporaryOrigin(
-        testOriginMetadata.mPersistenceType, principalInfo));
+        testOriginMetadata.mPersistenceType, principalInfo,
+        /* aCreateIfNonExistent */ true));
 
     {
       auto value =
@@ -1466,7 +1472,58 @@
 
   ASSERT_NO_FATAL_FAILURE(InitializeStorage());
   ASSERT_NO_FATAL_FAILURE(InitializeTemporaryStorage());
-  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryOrigin(GetTestOriginMetadata()));
+  ASSERT_NO_FATAL_FAILURE(
+      InitializeTemporaryOrigin(GetTestOriginMetadata(),
+                                /* aCreateIfNonExistent */ true));
+
+  ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
+  ASSERT_NO_FATAL_FAILURE(AssertTemporaryStorageInitialized());
+  ASSERT_NO_FATAL_FAILURE(
+      AssertTemporaryOriginInitialized(GetTestOriginMetadata()));
+
+  PerformOnBackgroundThread([]() {
+    auto testOriginMetadata = GetTestOriginMetadata();
+
+    nsCOMPtr<nsIPrincipal> principal =
+        BasePrincipal::CreateContentPrincipal(testOriginMetadata.mOrigin);
+    QM_TRY(MOZ_TO_RESULT(principal), QM_TEST_FAIL);
+
+    mozilla::ipc::PrincipalInfo principalInfo;
+    QM_TRY(MOZ_TO_RESULT(PrincipalToPrincipalInfo(principal, &principalInfo)),
+           QM_TEST_FAIL);
+
+    QuotaManager* quotaManager = QuotaManager::Get();
+    ASSERT_TRUE(quotaManager);
+
+    {
+      auto value = Await(quotaManager->ClearStoragesForOrigin(
+          /* aPersistenceType */ Nothing(), principalInfo));
+      ASSERT_TRUE(value.IsResolve());
+
+      ASSERT_TRUE(quotaManager->IsStorageInitialized());
+      ASSERT_TRUE(quotaManager->IsTemporaryStorageInitialized());
+      ASSERT_FALSE(quotaManager->IsTemporaryOriginInitialized(
+          testOriginMetadata.mPersistenceType, principalInfo));
+    }
+  });
+
+  ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
+
+  ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
+}
+
+TEST_F(TestQuotaManager, ClearStoragesForOrigin_NonExistentOriginDirectory) {
+  ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
+
+  ASSERT_NO_FATAL_FAILURE(AssertStorageNotInitialized());
+  ASSERT_NO_FATAL_FAILURE(AssertTemporaryStorageNotInitialized());
+  ASSERT_NO_FATAL_FAILURE(
+      AssertTemporaryOriginNotInitialized(GetTestOriginMetadata()));
+
+  ASSERT_NO_FATAL_FAILURE(InitializeStorage());
+  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryStorage());
+  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryOrigin(
+      GetTestOriginMetadata(), /* aCreateIfNonExistent */ false));
 
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
   ASSERT_NO_FATAL_FAILURE(AssertTemporaryStorageInitialized());
@@ -1515,7 +1572,58 @@
 
   ASSERT_NO_FATAL_FAILURE(InitializeStorage());
   ASSERT_NO_FATAL_FAILURE(InitializeTemporaryStorage());
-  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryOrigin(GetTestOriginMetadata()));
+  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryOrigin(
+      GetTestOriginMetadata(), /* aCreateIfNonExistent */ true));
+
+  ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
+  ASSERT_NO_FATAL_FAILURE(AssertTemporaryStorageInitialized());
+  ASSERT_NO_FATAL_FAILURE(
+      AssertTemporaryOriginInitialized(GetTestOriginMetadata()));
+
+  PerformOnBackgroundThread([]() {
+    auto testOriginMetadata = GetTestOriginMetadata();
+
+    nsCOMPtr<nsIPrincipal> principal =
+        BasePrincipal::CreateContentPrincipal(testOriginMetadata.mOrigin);
+    QM_TRY(MOZ_TO_RESULT(principal), QM_TEST_FAIL);
+
+    mozilla::ipc::PrincipalInfo principalInfo;
+    QM_TRY(MOZ_TO_RESULT(PrincipalToPrincipalInfo(principal, &principalInfo)),
+           QM_TEST_FAIL);
+
+    QuotaManager* quotaManager = QuotaManager::Get();
+    ASSERT_TRUE(quotaManager);
+
+    {
+      auto value = Await(quotaManager->ClearStoragesForOriginPrefix(
+          /* aPersistenceType */ Nothing(), principalInfo));
+      ASSERT_TRUE(value.IsResolve());
+
+      ASSERT_TRUE(quotaManager->IsStorageInitialized());
+      ASSERT_TRUE(quotaManager->IsTemporaryStorageInitialized());
+      ASSERT_FALSE(quotaManager->IsTemporaryOriginInitialized(
+          testOriginMetadata.mPersistenceType, principalInfo));
+    }
+  });
+
+  ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
+
+  ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
+}
+
+TEST_F(TestQuotaManager,
+       ClearStoragesForOriginPrefix_NonExistentOriginDirectory) {
+  ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
+
+  ASSERT_NO_FATAL_FAILURE(AssertStorageNotInitialized());
+  ASSERT_NO_FATAL_FAILURE(AssertTemporaryStorageNotInitialized());
+  ASSERT_NO_FATAL_FAILURE(
+      AssertTemporaryOriginNotInitialized(GetTestOriginMetadata()));
+
+  ASSERT_NO_FATAL_FAILURE(InitializeStorage());
+  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryStorage());
+  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryOrigin(
+      GetTestOriginMetadata(), /* aCreateIfNonExistent */ false));
 
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
   ASSERT_NO_FATAL_FAILURE(AssertTemporaryStorageInitialized());
@@ -1564,7 +1672,58 @@
 
   ASSERT_NO_FATAL_FAILURE(InitializeStorage());
   ASSERT_NO_FATAL_FAILURE(InitializeTemporaryStorage());
-  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryOrigin(GetTestOriginMetadata()));
+  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryOrigin(
+      GetTestOriginMetadata(), /* aCreateIfNonExistent */ true));
+
+  ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
+  ASSERT_NO_FATAL_FAILURE(AssertTemporaryStorageInitialized());
+  ASSERT_NO_FATAL_FAILURE(
+      AssertTemporaryOriginInitialized(GetTestOriginMetadata()));
+
+  PerformOnBackgroundThread([]() {
+    auto testOriginMetadata = GetTestOriginMetadata();
+
+    nsCOMPtr<nsIPrincipal> principal =
+        BasePrincipal::CreateContentPrincipal(testOriginMetadata.mOrigin);
+    QM_TRY(MOZ_TO_RESULT(principal), QM_TEST_FAIL);
+
+    mozilla::ipc::PrincipalInfo principalInfo;
+    QM_TRY(MOZ_TO_RESULT(PrincipalToPrincipalInfo(principal, &principalInfo)),
+           QM_TEST_FAIL);
+
+    QuotaManager* quotaManager = QuotaManager::Get();
+    ASSERT_TRUE(quotaManager);
+
+    {
+      auto value = Await(quotaManager->ClearStoragesForOriginAttributesPattern(
+          OriginAttributesPattern()));
+      ASSERT_TRUE(value.IsResolve());
+
+      ASSERT_TRUE(quotaManager->IsStorageInitialized());
+      ASSERT_TRUE(quotaManager->IsTemporaryStorageInitialized());
+      ASSERT_FALSE(quotaManager->IsTemporaryOriginInitialized(
+          testOriginMetadata.mPersistenceType, principalInfo));
+    }
+  });
+
+  ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
+
+  ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
+}
+
+TEST_F(TestQuotaManager,
+       ClearStoragesForOriginAttributesPattern_NonExistentOriginDirectory) {
+  ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
+
+  ASSERT_NO_FATAL_FAILURE(AssertStorageNotInitialized());
+  ASSERT_NO_FATAL_FAILURE(AssertTemporaryStorageNotInitialized());
+  ASSERT_NO_FATAL_FAILURE(
+      AssertTemporaryOriginNotInitialized(GetTestOriginMetadata()));
+
+  ASSERT_NO_FATAL_FAILURE(InitializeStorage());
+  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryStorage());
+  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryOrigin(
+      GetTestOriginMetadata(), /* aCreateIfNonExistent */ false));
 
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
   ASSERT_NO_FATAL_FAILURE(AssertTemporaryStorageInitialized());
@@ -1613,7 +1772,57 @@
 
   ASSERT_NO_FATAL_FAILURE(InitializeStorage());
   ASSERT_NO_FATAL_FAILURE(InitializeTemporaryStorage());
-  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryOrigin(GetTestOriginMetadata()));
+  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryOrigin(
+      GetTestOriginMetadata(), /* aCreateIfNonExistent */ true));
+
+  ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
+  ASSERT_NO_FATAL_FAILURE(AssertTemporaryStorageInitialized());
+  ASSERT_NO_FATAL_FAILURE(
+      AssertTemporaryOriginInitialized(GetTestOriginMetadata()));
+
+  PerformOnBackgroundThread([]() {
+    auto testOriginMetadata = GetTestOriginMetadata();
+
+    nsCOMPtr<nsIPrincipal> principal =
+        BasePrincipal::CreateContentPrincipal(testOriginMetadata.mOrigin);
+    QM_TRY(MOZ_TO_RESULT(principal), QM_TEST_FAIL);
+
+    mozilla::ipc::PrincipalInfo principalInfo;
+    QM_TRY(MOZ_TO_RESULT(PrincipalToPrincipalInfo(principal, &principalInfo)),
+           QM_TEST_FAIL);
+
+    QuotaManager* quotaManager = QuotaManager::Get();
+    ASSERT_TRUE(quotaManager);
+
+    {
+      auto value = Await(quotaManager->ShutdownStoragesForOrigin(
+          /* aPersistenceType */ Nothing(), principalInfo));
+      ASSERT_TRUE(value.IsResolve());
+
+      ASSERT_TRUE(quotaManager->IsStorageInitialized());
+      ASSERT_TRUE(quotaManager->IsTemporaryStorageInitialized());
+      ASSERT_FALSE(quotaManager->IsTemporaryOriginInitialized(
+          testOriginMetadata.mPersistenceType, principalInfo));
+    }
+  });
+
+  ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
+
+  ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
+}
+
+TEST_F(TestQuotaManager, ShutdownStoragesForOrigin_NonExistentOriginDirectory) {
+  ASSERT_NO_FATAL_FAILURE(ShutdownStorage());
+
+  ASSERT_NO_FATAL_FAILURE(AssertStorageNotInitialized());
+  ASSERT_NO_FATAL_FAILURE(AssertTemporaryStorageNotInitialized());
+  ASSERT_NO_FATAL_FAILURE(
+      AssertTemporaryOriginNotInitialized(GetTestOriginMetadata()));
+
+  ASSERT_NO_FATAL_FAILURE(InitializeStorage());
+  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryStorage());
+  ASSERT_NO_FATAL_FAILURE(InitializeTemporaryOrigin(
+      GetTestOriginMetadata(), /* aCreateIfNonExistent */ false));
 
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
   ASSERT_NO_FATAL_FAILURE(AssertTemporaryStorageInitialized());